lib/raty.js

```diff
@@ -0,0 +1,784 @@
+        }
+      });
+    });
+  }
+
+  // TODO: model spec
+  _bindOverCancel() {
+    var that = this;
+
+    this.cancelButton.addEventListener('mouseover', function (evt) {
+      if (that.self.dataset.readOnly && !that.self.dataset.readOnly.toBoolean()) {
+        return;
+      }
+
+      var starOff = that.opt.path + that.opt.starOff;
+      var icon = that.opt.cancelOn;
+
+      if (that.opt.starType === 'img') {
+        that.stars.forEach((value) => {
+          value.src = starOff;
+        });
+      } else {
+        icon = that.opt.cancelClass + ' ' + icon;
+        that.stars.forEach((value) => {
+          value.className = starOff;
+        });
+      }
+
+      that._setIcon(that.cancelButton, icon);
+      that._target(null, evt);
+
+      if (that.opt.mouseover) {
+        that.opt.mouseover.call(that.self, null);
+      }
+    });
+  }
+
+  // TODO: model spec
+  _buildScoreField() {
+    const input = document.createElement('input');
+    input.name = this.opt.scoreName;
+    input.type = 'hidden';
+    this.self.appendChild(input);
+    return input;
+  }
+
+  // TODO: model spec
+  _createCancel() {
+    let icon = this.opt.path + this.opt.cancelOff;
+    let button = document.createElement(this.opt.starType);
+    const attributes = {
+      title: this.opt.cancelHint,
+      class: this.opt.cancelClass,
+    };
+
+    if (this.opt.starType === 'img') {
+      attributes.src = icon;
+      attributes.alt = 'x';
+    } else {
+      attributes['data-alt'] = 'x';
+      attributes.class = icon;
+    }
+    for (const key in attributes) {
+      button.setAttribute(key, attributes[key]);
+    }
+    if (this.opt.cancelPlace === 'left') {
+      this.self.prepend('\u00A0');
+      this.self.prepend(button);
+    } else {
+      this.self.append('\u00A0');
+      this.self.appendChild(button);
+    }
+    button.id = 'raty-cancel-btn';
+    this.cancelButton = document.querySelector('#raty-cancel-btn');
+  }
+
+  // TODO: model spec
+  _createScore() {
+    var score = document.querySelector(this.opt.targetScore);
+
+    this.scoreField = score ?? this._buildScoreField();
+  }
+
+  _createStars() {
+    for (var i = 1; i <= this.opt.number; i++) {
+      var attributes = this._attributesForIndex(i);
+      const img = document.createElement("img");
+      const element = document.createElement(this.opt.starType);
+      let el = this.opt.starType === 'img' ? img : element;
+      el.id = 'raty-rate-' + i;
+
+      for (const key in attributes) {
+        el.setAttribute(key, attributes[key]);
+      }
+
+      this.self.appendChild(el);
+
+      if (this.opt.space && i < this.opt.number) {
+        this.self.append('\u00A0');
+      }
+    }
+
+    let stars = document.querySelectorAll(`${this.element} > ${this.opt.starType}[id^="raty-rate"]`);
+
+    if (this.cancelButton) {
+      stars.filter((value) => value.id !== '#raty-cancel-btn');
+    }
+    console.log("hii");
+    this.stars = stars;
```

Comment:
If the attribute was added for filter purpose, after get it, we should remove it.

Reply:
no worries _createStars will move up to the _createCancel